### ISSUE TITLE
Certain conv tests failing on METAL with TC=1

### ISSUE THREAD

**Issue by ZenginU** [tinygrad/tinygrad#1302]
On up to date forked master branch, some of the tests are failing, `test_conv/test_first_three, test_optim/test_adam_high_lr, test_optim/test_adamw_high_lr`. Is this normal to fail in local machines (Macos METAL)? I just run the test folder in pycharm without any paramaters.

For the test_first_three, problem happens to be shift operation in [linearizer.py/hand_coded_optimizations](https://github.com/tinygrad/tinygrad/blob/8562b5a04f2fd2e39b7ddc0578abc2dfe377f774/tinygrad/codegen/linearizer.py#L666C9-L666C62). It tries to shift 4 with 8 which creates [0, 8] shape that cant be reshaped.

**Comment by wozeparrot** [comment]
iirc the high_lr tests have some issues on metal.

**Comment by ZenginU** [comment]
What about test_conv/test_first_three?

**Comment by wozeparrot** [comment]
Yea, I haven't heard of that failing before. Can you bisect?

**Comment by ZenginU** [comment]
What you mean by can you bisect ? Sorry, english is not my first language. Do you mean open another issue for test_conv/test_first_three ?

**Comment by wozeparrot** [comment]
Like a git bisect to find the bad commit.

**Comment by ZenginU** [comment]
Learned new thing today, I believe its this commit #1212. But it seems it is not related

**Comment by bogdad** [comment]
For me,
`env METAL=1 TC=1 PYTHONPATH=. python3 test/test_conv.py -k "test_first_three"`
Gets an `AssertionError: shape must be symbolic ints and can't contain 0 or negative numbers (16, 8192, 0, 8, 8)`
(Also fails for `env CUDA=1 TC=1 PYTHONPATH=.` in the same way)

I think, that is because in hand_coded_optimizations() we are choosing s0 and s1 by looking into axis of buf0 and buf1 having zero stride.
If buf0 or buf1 are Tensor.ones(…), it will have all strides being 0 due to expand() not allocating anything new.
It is happens that in tests, ie test_conv test_first_three, k.first_reduce == 2, and s0 == 0 and s1 == 2.
Then we are doing `k.shift_to(k.first_reduce, 8)`which transforms one of the shape trackers (16, 8192, 32) to (16, 8192, 4, 8).
Next, `k.shift_to(s1, 8, insert_before=k.first_reduce)` tries to shift (16, 8192, 4, 8) by 8 on axis 2, and gets (4//8, 8) which leads to assertion above.

I hacked it making s0,s1,k.first_reduce always unique, and it makes the tests pass,
```
       # TODO: select axis in smart way
-      s0, s1 = axis_buf0[-1][0], axis_buf1[-1][0]
+      s0 = next((s[0] for s in reversed(axis_buf0) if s[0] != k.first_reduce), None)
+      s1 = next((s[0] for s in reversed(axis_buf1) if s[0] != k.first_reduce and s[0] != s0), None)
+

vladimir@FT751F6N7D ~/w/tinygrad (hand_coded_optimizations_tensor_cores_make_s0_s1_first_reduce_unique)> env METAL=1 TC=1 PYTHONPATH=. python3 test/test_conv.py -k "test_first_three"
(1, 16, 64, 128)
----------------------------------------------------------------------
Ran 1 test in 0.174s
OK
```

I don’t understand the the series of shifts_to in hand_coded_optimizations() for tensor cores yet, maybe it is fine to have some of s0, s1 and k.first_reduce equal as long as they are of proper sizes.

**Comment by ZenginU** [comment]
Fixed by #1582
